### ISSUE TITLE
Add support for real(32) params

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#5122]
This issue relates to test/param/errors/param32BitFloat.future which was added in PR#5072.

It seems that we don't / can't support real(32) params today due to the fact that the _defaultOf() routines can't be written in a way that returns a param 32-bit 0.0 value.  In user code, we'd typically write "0.0:real(32)" but we can't do this cast at param-evaluation time because we have traditionally not done floating point operations at compile-time (for fear that the floating point semantics might differ at compile-time from what the user would get at execution time -- perhaps this is being overly paranoid, though, since the user did request us to do the computation at compile-time.... but that's a separate issue).

This seems like it's probably just a bootstrapping problem and could potentially be handled by special-casing this value using a primitive or somesuch.

**Comment by nimitbhardwaj** [comment]
Hey, I am encountering a bug here, its related to the behaviour of real(32).

What I did was simply

```chapel
var a:real(32) = -1.5;
writeln(a);
```
Its giving error `a.chpl:1: error: type mismatch in assignment from real(64) to real(32)`
while for the code
```chapel
var a:real(32) = 1.5;
writeln(a);
```
there is no error, smooth compilation, there is an error related to parsing of negative values which are real(32), in my sense real(32) is a float in C or C++, so there are 11 bits for decimal part and one bit for sign, it should work, -1.5 is also a real(32) or float, why this is, am I wrong ??

One more thing, it works correctly for

```chapel
var a:real(32) = -100;
writeln(a);
```
This above program is compiled correctly, only negatives with a decimal provides error

**Comment by bradcray** [comment]
The issue in your code is that on master today:
* we do permit coercions from `int(64)`s (like `-100`) to `real(32)`s when there is no loss of information (@mppf: did I get that right?), but
* we don't permit coercions from `real(64)`s (like `-1.0`) to `real(32)`s (ever).
The workaround for the time being is to do an explicit downcast, like: `(-1.0):real(32)`

We have discussed adding support for automatic downcasting from `real(64)` values to `real(32)` values in the event that there would be no loss of information, but that has not yet been implemented.


> there is an error related to parsing of negative values

The parsing of the values is not the problem here, it's the automatic conversion from a more precise floating point types to less-precise ones.

Not related to your question, but as a side note:  in most languages, including Chapel, there aren't negative literal values, there are only positive negative values to which the unary negative operator can be applied.  Thus `-1` isn't parsed as "the integer with the value `-1`" but rather "the unary negative operator applied to the positive integer 1."

**Comment by bradcray** [comment]
Noting belatedly that this issue is also roughly related to #7692 and #5922.

**Comment by mppf** [comment]
@bradcray - actually I did implement it in this (wip/closed) PR #7668. We just need to decide it is "OK" to do param real/real(32) operations before I can proceed with it.

**Comment by bradcray** [comment]
@mppf: Any chance that your PR #12386 can close (or help close) this one as well?

**Comment by mppf** [comment]
@bradcray - yes, thanks. This issue is resolved by PR #12386. That PR un-futures param32BitFloat.chpl.

I also verified that this test (brought up in the discussion on this issue) now compiles and runs:
``` chapel
var a:real(32) = -1.5;
writeln(a);
```
